library/core/src/fmt/mod.rs

```diff
@@ -2281,16 +2282,50 @@ impl<T: ?Sized> Pointer for &mut T {
 
 // Implementation of Display/Debug for various core types
 
+/// A local trait for pointer Debug impl so that we can have
+/// min_specialization-compliant specializations for two types of fat ptrs.
+trait PtrMetadataFmt {
+    fn fmt_ptr(&self, ptr: *const (), f: &mut Formatter<'_>) -> Result;
+}
+
+// Regular pointer / default impl
+impl<T> PtrMetadataFmt for T {
+    #[inline]
+    default fn fmt_ptr(&self, ptr: *const (), f: &mut Formatter<'_>) -> Result {
+        Pointer::fmt(&ptr, f)
+    }
+}
+
+// Pointer + length
+impl PtrMetadataFmt for usize {
+    #[inline]
+    fn fmt_ptr(&self, ptr: *const (), f: &mut Formatter<'_>) -> Result {
+        f.write_fmt(format_args!("({:p}, {})", ptr, *self))
+    }
+}
+
+// Pointer + vtable
+impl<Dyn: ?Sized> PtrMetadataFmt for DynMetadata<Dyn> {
+    #[inline]
+    fn fmt_ptr(&self, ptr: *const (), f: &mut Formatter<'_>) -> Result {
+        f.debug_tuple("")
+            .field(&ptr)
+            .field(self)
+            .finish()
+   }
+}
+
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> Debug for *const T {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        Pointer::fmt(self, f)
+        let meta = ptr::metadata(*self);
+        meta.fmt_ptr(*self as *const (), f)
     }
 }
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> Debug for *mut T {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        Pointer::fmt(self, f)
+        Debug::fmt(&(*self as *const _), f)
```

Comment:
Combinations of `&` `*` and `as` quickly become hard to read. Maybe split this in two lines and use implicit coercion rather than `as`:

```suggestion
        let ptr: *const T = *self;
        Debug::fmt(&ptr, f)
```